ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/TerminateQueryExecutor.java

```
@@ -0,0 +1,83 @@
+import io.confluent.ksql.util.KsqlConstants.KsqlQueryStatus;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public final class TerminateQueryExecutor {
+
+  private TerminateQueryExecutor() { }
+
+  public static Optional<KsqlEntity> execute(
+      final ConfiguredStatement<TerminateQuery> statement,
+      final SessionProperties sessionProperties,
+      final KsqlExecutionContext executionContext,
+      final ServiceContext serviceContext,
+      final DistributingExecutor distributingExecutor,
+      final KsqlSecurityContext securityContext
+  ) {
+    final TerminateQuery terminateQuery = statement.getStatement();
+    final Optional<QueryId> queryId = terminateQuery.getQueryId();
+    final RemoteHostExecutor remoteHostExecutor = RemoteHostExecutor.create(
+        statement,
+        sessionProperties,
+        executionContext,
+        serviceContext.getKsqlClient()
+    );
+
+    if (executionContext.getPersistentQuery(queryId.get()).isPresent()) {
+      // do default behaviour for terminating persistent queries
+      distributingExecutor.execute(statement, executionContext, securityContext);
+    } else {
+      // Check are we running this push query locally, if yes then terminate, otherwise
+      // propagate terminate query to other nodes
+      if (executionContext.getQuery(queryId.get()).isPresent()) {
+        executionContext.getQuery(queryId.get()).get().close();
+        return Optional.of(
+            new TerminateQueryEntity(statement.getStatementText(), queryId.get().toString(), true)
+        );
+      } else {
+        final boolean wasTerminatedLocally = remoteHostExecutor.fetchAllRemoteResults().getLeft()
+            .values()
+            .stream()
+            .map(TerminateQueryEntity.class::cast)
+            .map(TerminateQueryEntity::getWasTerminatedLocally)
+            .anyMatch(b -> b.equals(true));
+        return Optional.of(new TerminateQueryEntity(
```

if not wasTerminatedRemotely, we should probably throw an error because we expect it to happen somewhere.